### PR TITLE
Set depth in `VolumeVisual` for `mip`/`minip`/`attenuated_mip` rendering modes

### PR DESCRIPTION
**Pull request by alisterburt**
This PR sets the depth buffer correctly in the `mip`/`minip`/`attenuated_mip` rendering modes of the `VolumeVisual` - if in plane rendering mode the depth is always set to the ray-plane intersection as before (#2289)

https://user-images.githubusercontent.com/7307488/156226184-e0146fd2-a0af-4e02-8148-1aa8c6af5200.mp4

This allows us to compose multiple volumes in the same space in opaque blending modes (excuse the lack of alpha in the cmap)

https://user-images.githubusercontent.com/7307488/156226373-7dbd943d-5be9-4dd1-b5e8-23927e6e6e64.mp4

cc @brisvag

### CODE REVIEW

**Review by djhoese — CHANGES_REQUESTED** [review]
Cool! Tests please!

I'm also not sure I'm the right person to give a full review of the implementation so @almarklein please review if you have the time.

**Comment by alisterburt** [comment]
Thanks! Will follow up with a test 🙂 

Some implementation notes which may be useful when reviewing
- I make use of the depth calculation mechanism introduced by @kevinyamauchi in #2190 
- the depth buffer is already set for any fragments not discarded to `gl_FragCoord.z` so overall impact on speed should be minimal
https://github.com/vispy/vispy/blob/bf793b36010d3cecb8307e1360f6549633fed1e5/vispy/visuals/volume.py#L294-L296
- mip/minip may be slightly slower - during the 'after_loop' an if statement was needed to keep track of the position where the new maximum value occurred

File: vispy/visuals/volume.py [hunk 1]
```diff
@@ -361,16 +369,18 @@
     vec3 step = N / u_shape;
     vec3 start_loc = intersection_tex - ((step * f_nsteps) / 2);
 
-    // Set depth value
-    surface_point = intersection;
-    surface_found = true;
+    // Ensure that frag depth value will be set to plane intersection
+    rendering_as_plane = true;
+    set_frag_depth = true;
+    frag_depth_point = intersection;
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply vispy/visuals/volume.py:375]: `frag_depth_point` is overriden by some of the snippets, I should remove it up here though :)

File: vispy/visuals/volume.py [hunk 2]
```diff
@@ -509,7 +543,7 @@
         """,
     after_loop="""
 
-        if (!surface_found) {
+        if (!set_frag_depth) {
```
[the remark this answers is not in the document]
Comment (alisterburt) [reply vispy/visuals/volume.py:546]: This flag is used in other rendering modes to decide whether to set the depth, iso is a special case where we discard fragments if no surface is found - I can leave `surface_found` and `set_frag_depth` for iso if you prefer but can't switch to `keep_fragment` across all rendering modes

**Comment by alisterburt** [comment]
thanks for the review @brisvag - I don't know what you're referring to with the clipping planes but I can make an example with transforms

File: vispy/visuals/volume.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (brisvag) [reply vispy/visuals/volume.py:546]: Makes sense! Then leaving as is seems fine to me :)

**Comment by brisvag** [comment]
It was just an example of a time where testing with just a cubic, untransformed volume came back to bite us later :P I think a test with anisotropic shape should be enough here actually, but you never know.

**Comment by alisterburt** [comment]
I hope this test works but I'm not sure it will... here's what I see on screen for the example
image

but in the result of `c.render()` I see all blue, I'm not sure why. Do you have an idea about why these might be different @djhoese ?

**Comment by djhoese** [comment]
I'm not sure I understand the asserts at the end. The comment says render with grays? The earlier comment says we should only see blue? But then the asserts check for reds and blues?

**Comment by alisterburt** [comment]
sorry for the confusion, comment was copy and pasted from Lorenzo's plane example in the same ile, I'm not sure whether it's correct there either

The example and what I see on screen is blue surrounded by red. When I look at the array produced by `c.render()` I only see blue so I don't think the test is working properly - I'm not sure where the discrepancy between the array generated by `c.render()` and what I see on screen comes from 🙁

File: vispy/visuals/tests/test_volume.py [hunk 3]
```diff
@@ -333,4 +333,55 @@ def test_plane_depth():
...
+            parent=v.scene,
+        )
+
+        # render
+        rendered = c.render()
+        reds = np.sum(rendered[:, :, 0])
+        greens = np.sum(rendered[:, :, 1])
+        blues = np.sum(rendered[:, :, 2])
+        assert reds > 0
+        assert np.allclose(greens, 0)
```
Comment (djhoese) [inline vispy/visuals/tests/test_volume.py:383]:
```suggestion
        np.testing.assert_allclose(greens, 0)
```

**Comment by djhoese** [comment]
So I'm guessing you converted the test function to a separate script? Or how did you get what you said you see on screen?

**Comment by alisterburt** [comment]
not quite, I drop into a debugger in the test itself and push the rendered array into napari for visualisation - when I do this, the testing canvas from vispy also becomes visible

**Comment by djhoese** [comment]
Just to be sure I copied the test code to a script and did a few modifications. The canvas window shows all blue.

```python
from vispy import app
from vispy import scene
import numpy as np

c = scene.SceneCanvas(size=(40, 40))
v = c.central_widget.add_view(border_width=0)
v.camera = 'arcball'
v.camera.fov = 0
v.camera.center = (20, 20, 20)
v.camera.scale_factor = 40.0

# A blue strip behind a red strip
# If depth is set correctly, we should see only red pixels
# the screen
blue_vol = np.zeros((40, 40, 40), dtype=np.uint8)
blue_vol[:, 39, :] = 1  # back plane blue
blue_vol[15:25, 15:25, 15:25] = 1  # blue in center

red_vol = np.zeros((40, 40, 40), dtype=np.uint8)
red_vol[:, 35, :] = 1  # red plane in front of blue plane

scene.visuals.Volume(
    blue_vol,
    interpolation="nearest",
    clim=(0, 1),
    cmap="blues",
    parent=v.scene,
)

scene.visuals.Volume(
    red_vol,
    interpolation="nearest",
    clim=(0, 1),
    cmap="reds",
    parent=v.scene,
)

# render
# rendered = c.render()
# reds = np.sum(rendered[:, :, 0])
# greens = np.sum(rendered[:, :, 1])
# blues = np.sum(rendered[:, :, 2])
# assert reds > 0
# assert np.allclose(greens, 0)
# assert blues > 0
c.show()
app.run()
```

**Comment by djhoese** [comment]
If I expand the window and turn the camera, here's what I see:

image

The blue face on the left was what I saw when the window was first shown.

**Comment by alisterburt** [comment]
running your example script as is on this branch I see 
image

If I try to render the canvas in this context I get a bunch of ugly errors

```python
WARNING: Error drawing visual <Volume at 0x7fb9f8e01e80>
WARNING: Traceback (most recent call last):
  File "/Users/aburt/Programming/vispy/examples/vispy_test.py", line 46, in <module>
    c.show()
  File "/Users/aburt/Programming/vispy/vispy/app/canvas.py", line 435, in show
    self._backend._vispy_set_visible(visible)
  File "/Users/aburt/Programming/vispy/vispy/app/backends/_qt.py", line 471, in _vispy_set_visible
    self.showNormal()
  File "/Users/aburt/Programming/vispy/vispy/app/backends/_qt.py", line 567, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/aburt/Programming/vispy/vispy/app/backends/_qt.py", line 567, in event
    out = super(QtBaseCanvasBackend, self).event(ev)
  File "/Users/aburt/Programming/vispy/vispy/app/backends/_qt.py", line 903, in paintGL
    self._vispy_canvas.events.draw(region=None)
  File "/Users/aburt/Programming/vispy/vispy/util/event.py", line 453, in __call__
    self._invoke_callback(cb, event)
  File "/Users/aburt/Programming/vispy/vispy/util/event.py", line 471, in _invoke_callback
    _handle_exception(self.ignore_callback_errors,
  << caught exception here: >>
  File "/Users/aburt/Programming/vispy/vispy/util/event.py", line 469, in _invoke_callback
    cb(event)
  File "/Users/aburt/Programming/vispy/vispy/scene/canvas.py", line 218, in on_draw
    self._draw_scene()
  File "/Users/aburt/Programming/vispy/vispy/scene/canvas.py", line 277, in _draw_scene
    self.draw_visual(self.scene)
  File "/Users/aburt/Programming/vispy/vispy/scene/canvas.py", line 315, in draw_visual
    node.draw()
  File "/Users/aburt/Programming/vispy/vispy/scene/visuals.py", line 103, in draw
    self._visual_superclass.draw(self)
  File "/Users/aburt/Programming/vispy/vispy/visuals/visual.py", line 451, in draw
    self._program.draw(self._vshare.draw_mode,
  File "/Users/aburt/Programming/vispy/vispy/visuals/shaders/program.py", line 102, in draw
    Program.draw(self, *args, **kwargs)
  File "/Users/aburt/Programming/vispy/vispy/gloo/program.py", line 526, in draw
    canvas.context.flush_commands()
  File "/Users/aburt/Programming/vispy/vispy/gloo/context.py", line 172, in flush_commands
    self.glir.flush(self.shared.parser)
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 579, in flush
    self._shared.flush(parser)
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 501, in flush
    parser.parse(self._filter(self.clear(), parser))
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 819, in parse
    self._parse(command)
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 795, in _parse
    ob.set_framebuffer(*args)
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 1739, in set_framebuffer
    self.deactivate()
  File "/Users/aburt/Programming/vispy/vispy/gloo/glir.py", line 1753, in deactivate
    gl.glBindFramebuffer(gl.GL_FRAMEBUFFER, stack[-1])
IndexError: list index out of range
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt5) at 0x7fba082067c0>> for DrawEvent
WARNING: Error drawing visual <Volume at 0x7fb9f8e01e80>
ERROR: Invoking <bound method SceneCanvas.on_draw of <SceneCanvas (PyQt5) at 0x7fba082067c0>> repeat 2
WARNING: Error drawing visual <Volume at 0x7fb9f8e01e80>
```

I'm not sure why you would not see the scene correctly and I would, that's definitely concerning - @brisvag could you give it a quick run and report what you see?

The strange effects on rotating are somewhat expected I think, some of the rays will trace the full volume without finding any non-zero values so the first zero found on the front face is the max value. The depth is then set to the front face

Sorry this couldn't be a simple one 😆 

To summarise
- scene appears to be setup correctly
- I see depth rendered properly locally (blue on red bg)
- You see depth rendered wrong locally (all blue)
- rendering from the canvas in the test appears wrong (all blue)

I'm not sure where this leaves us exactly, sorry again!

**Comment by djhoese** [comment]
If you pull your window bigger and rotate it, what does it look like? I need an idea of what the overall combination of volumes looks like on your machine.

**Comment by alisterburt** [comment]
tadaa

image

**Comment by djhoese** [comment]
Interesting. For the record I'm on a PopOS (think Ubuntu) system with Nvidia GPU and I think I ran the example with Python 3.7.

**Comment by alisterburt** [comment]
Ooh, another pop! fan! 

here is my info (from `napari --info` installed in the same env)

Platform: macOS-10.16-x86_64-i386-64bit
System: MacOS 11.6
Python: 3.9.6 (default, Jul 30 2021, 09:31:09)  [Clang 10.0.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
VisPy: 0.9.5.dev3+dirty

OpenGL:
  - GL version:  2.1 Metal - 71.7.1
  - MAX_TEXTURE_SIZE: 16384

**Comment by alisterburt** [comment]
heading off and won't be able to push on this again for a couple days at least, just FYI - thanks for the help so far!

**Comment by brisvag** [comment]
Well, I see the same a @djhoese... the plot thickens!

**Comment by brisvag** [comment]
> tadaa
> 
> image

This is definitely fishy; the cube in the center also looks deformed.

**Comment by alisterburt** [comment]
it's a depth issue from the zero values elsewhere being max along certain rays - as long as the cube is in front of the plane (how the example was setup) everything is as I expect

image

**Comment by brisvag** [comment]
I'm not convinced that things look correct in yours either. I would expect the red one to be a slab that goes to the edges of the cube, but it looks like a plane that somehow ends somwhere within the cube. Also, the area of white between the red and the blue slab should be blue, not white, afaik. 

Still doesn't explain why whe would see different things, though.

**Comment by alisterburt** [comment]
yeah agree that I have no idea why we see differences

You see white there because the red plane volume has only zeros along those rays, max is on front face and in front of blue

I don't know why the max values are rendering white though 🤣

**Comment by djhoese** [comment]
@brisvag remind me of your OS? Windows? Mac? Linux?

**Comment by alisterburt** [comment]
@brisvag is an arch Linux enthusiast

**Comment by djhoese** [comment]
And why is white supposed to be transparent/removed? Isn't it a valid color in the image/volume?

**Comment by brisvag** [comment]
Hah! Found it! `max_loc_tex` was not initialized to `start_loc_refine`! So my new best guess for the different behaviour is that our pc somehow have different policies for initializing values on the GPU?

**Comment by brisvag** [comment]
> And why is white supposed to be transparent/removed? Isn't it a valid color in the image/volume?

It's not "supposed" to be removed, necessarily. It depends on your colormap. Do you consider a value of zero to be "no data" or "data value of zero"? Either way, this is a useful tool to have, and we can leave the default as they are currently on main to not disrupt anything.

**Comment by alisterburt** [comment]
The white is basically a rendering artefact here - I would have assumed that 0 would map to black rather than white but removing it completely is prettier and gives a better sense of what is going on

Wow @brisvag ! I have no idea how that would still work on my machine but so glad you found it

**Comment by djhoese** [comment]
I don't look at volumes much, but in an image I would expect the colormap to be respected (0 -> white in this case). In a Volume if there is a convention that all zeros are transparent then I'm OK with it being removed. Whatever makes sense to you guys...with a sanity check from @almarklein maybe.

**Comment by brisvag** [comment]
I don't know about conventions... I'm just thinking of what I would want to see; the extreme case is where you have an image of full zeros that's above other data in the scenegraph. Without this trick, you'd just see white everywhere, cause you always encounter a white pixel there and it would win over the other data.

But again, I would not make this default.

**Comment by brisvag** [comment]
Ok, so this is how it works now: first is the default setup (current main), then I set the `mip_cutoff` for each volume to `0`.


https://user-images.githubusercontent.com/23482191/156793739-07ba45da-4ce7-4ac5-99ce-59d2eb7f351f.mp4

**Comment by brisvag** [comment]
Well, this sucks, the tests that fail here all pass locally.

**Comment by brisvag** [comment]
Ok, I dug a bit more into the test failure. I can reproduce them locally, but they make no sense. If I slowly make my way through the test code, I see this:
image
which is perfectly fine.
Then, when I do `reds = c.render()[:, :, 0].sum()` I get `0`. Clearly, `SceneCanvas.render` is broken.
If I remove the blue volume (by simply not assigning its parent), then we DO see the red as expected.

Something in `SceneCanvas.render` is messing up with depth testing. Indeed, if I simply switch the order of the two volumes in the scenegraph, the test is fixed, at least locally.

I think this needs to be fixed separately and shouldn't hold this PR.

**Comment by djhoese** [comment]
So I ran the example with the current version of the branch and let me see if I can dumb down the issue to make sure I understand it. The test script when I run it shows a blue square in the middle of a red square. The 3D space surrounding the blue square/cube is actually a "white" cube. When you do `.render()` you're getting the value of the white cube rather than the red plane behind it. Right?

Now while blaming `.render()` is the easy solution here, I'm not sure I want to merge this without knowing the `.render()` issue first. Especially considering since `.render()` is relatively simple and isn't the thing that changed here.

**Comment by brisvag** [comment]
> The test script when I run it shows a blue square in the middle of a red square. The 3D space surrounding the blue square/cube is actually a "white" cube. When you do .render() you're getting the value of the white cube rather than the red plane behind it. Right?

No, cause white would be `(255, 255, 255, 255)`, right? We see `(0, 0, 0, 255)`, which is odd.

I think the most important thing is that this is on the same canvas that I'm looking at on my screen where the rendering is correct. Clearly the way things are rendered to screen and the way they are rendered offscreen is different.

> Now while blaming .render() is the easy solution here, I'm not sure I want to merge this without knowing the .render() issue first. Especially considering since .render() is relatively simple and isn't the thing that changed here.

This might just be the first time that the effects of the depth buffer got tested with multiple visuals that overlap *just right*. I don't see a reason why the code in this PR could be the cause of that, seeing that it's only shader code...

> Also if you look at the other channels of the .render() it looks like the result is almost all blue. That means it is seeing the back volume/plane, right?

Yes; once @alisterburt merges my change (simply swap the order of the senegraph) you'll see that (hopefully) everything works, so `mip` and `depth` are calculated correctly (if they didn't we would either see only blue or only red, at most mixed with white).

**Comment by djhoese** [comment]
With the PR changes, this doesn't look right:

image

Is the blue part of the volume supposed to be cut off by the white?

**Comment by alisterburt** [comment]
the example is only set up to render correctly when viewed face on - as long as you see blue on red in that case then everything is working as expected, depth is being set appropriately for regions which are not 'all zero' for the ray cast through the volume :)

File: vispy/visuals/volume.py [hunk 4]
```diff
@@ -268,36 +267,27 @@
...
             }
             // Advance location deeper into the volume
             loc += step;
         }
     }
-    
-    // discard fragment if texture not sampled
-    if ( texture_sampled != 1 ) {
+
+    if ( texture_sampled == false ) {
```
[the remark this answers is not in the document]
Comment (brisvag) [reply vispy/visuals/volume.py:279]: Oh, glsl is more advanced than I though :P

File: vispy/visuals/volume.py [hunk 5]
```diff
@@ -497,23 +525,19 @@
                     gl_FragColor = applyColormap(color.r);
 
                     // set the variables for the depth buffer
-                    surface_point = iloc * u_shape;
-                    surface_found = true;
+                    frag_depth_point = iloc * u_shape;
```
[the remark this answers is not in the document]
Comment (brisvag) [reply vispy/visuals/volume.py:528]: Instead of juggling around the variable, I made it more explicit: the `discard `happens below in the `else` block. Though I realize now that maybe it should be one level deeper, cause we want to be able to discard fragments that passed the first coarse test?

[the remark this answers is not in the document]
Comment (almarklein) [reply vispy/visuals/volume.py:528]: Oh wait, this is the refinement stage. In that case I think its fine, because I don't think its possible/likely to hit the threshold in the coarse test and not in the refinement.

[the remark this answers is not in the document]
Comment (brisvag) [reply vispy/visuals/volume.py:528]: I guess it can happen if you pass by parallel to a surface from within 0.2 distrance...

[the remark this answers is not in the document]
Comment (almarklein) [reply vispy/visuals/volume.py:528]: In that case the values for the coarse location would be used, right?

[the remark this answers is not in the document]
Comment (brisvag) [reply vispy/visuals/volume.py:528]: Mhm... In that case the fragment would not be discarded, and it's depth would stay at the default, which is at the closest face of the cube. Also, it's color would be unset, and thus transparent. This can be a problem with transparency issues, so I think we should solve it.

Wait, actually it's all wrong here, cause I'm discarding *in the loop*. Thius should happen *after the loop* anyways o.o

**Comment by brisvag** [comment]
Ah I see why I got confused: the other snippets to the refinement in the `after_loop` instead.

**Comment by brisvag** [comment]
Made a PR with some fixes, @alisterburt alisterburt/vispy#4

**Comment by brisvag** [comment]
> Made a PR with some fixes, @alisterburt [alisterburt#4](https://github.com/alisterburt/vispy/pull/4)

PR updated with fixes from my findings at #2308!

**Review by almarklein — APPROVED** [review]

**Comment by djhoese** [comment]
Sweet. Let's merge this and then we can talk about the other PR. Thanks @alisterburt and @brisvag for working on all of this and finding the weird Qt behavior. I'm not sure I follow everything in this PR and the other one so let me know if there is something else that needs to be done before merging the other PR.

**Comment by alisterburt** [comment]
Thanks for your time during review @djhoese and @almarklein 🙂

**Comment by brisvag** [comment]
Thanks everyone for the efforts! I feel like we turned a corner with these stubborn tests, and future PRs should be easier :)

**Comment by Wangqkoe** [comment]
> This PR sets the depth buffer correctly in the `mip`/`minip`/`attenuated_mip` rendering modes of the `VolumeVisual` - if in plane rendering mode the depth is always set to the ray-plane intersection as before (#2289)
> 
>  composable-vispy-depth-volume-plane.mp4 
> This allows us to compose multiple volumes in the same space in opaque blending modes (excuse the lack of alpha in the cmap)
> 
>  cells3d-composed.mp4 
> cc @brisvag

Hi, alisterburt. I'm currently trying to plot some voxel figures using napari and insert a plane image in a specific z-axis position. But the existing rending mode in napari is not capable of realizing the effect that I want until I saw your demo here, which is exactly what I want to achieve. Could you tell me how can I get your version of napari？ 😊

**Comment by alisterburt** [comment]
Hi @Wangqkoe - super cool! Plane rendering itself is supported in napari>=0.4.15 and has GUI controls, see https://github.com/napari/napari/blob/main/examples/volume_plane_rendering.py 

Please change the blending mode to 'translucent' if trying to overlay multiple volumes with depth, to get depth working properly it should be as simple as installing the latest version of vispy from main in whatever environment you're using to run napari, let me know if you need a bit more direction 

One thing you will have to be mindful of is that the default colormaps in napari will render low intensity as black, you will probably need a custom colormap with some transparency below a certain threshold for the effect to look how you hope :) see some discussion in napari/napari#3575

I would be excited to see the figure when you get things working! Don't hesitate to reach out 🙂

**Comment by brisvag** [comment]
To add to what @alisterburt said, there's also #2308 which was designed to fix precisely that end-of-the-colormap issue; the plan is to hook that up in napari as soon as the latest changes in vipy get released :)
